Scrollbar doesn't update on CSS color-scheme change

lukewarlow opened the pull request:
#### 044d44c8940b219fb3d4f65e0cda8f97a4f86f30
<pre>
Scrollbar doesn&apos;t update on CSS color-scheme change
<a href="https://bugs.webkit.org/show_bug.cgi?id=244494">https://bugs.webkit.org/show_bug.cgi?id=244494</a>

Reviewed by Aditya Keerthi.

The color-scheme property is now diffed and triggers
invalidateScrollbars on scrollable areas.

invalidateScrollbars also now calls invalidateScrollCorners.

This causes the scrollbar and scroll corner to correctly rerender without
hovering or invalidating through other means. When UI Side Compositing is
disabled.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-1-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-1-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-1.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-2-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-2-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-2.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-3-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-3-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-3.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-4-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-4-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-4.html: Added.
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::invalidateScrollbars):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::styleDidChange):

Canonical link: <a href="https://commits.webkit.org/266176@main">https://commits.webkit.org/266176@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/23c60ef0dd6661c2c699b4914e631171daf5fbfe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13118 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13432 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13765 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14856 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12499 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13186 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15940 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13459 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15194 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13285 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11092 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15343 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11257 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11847 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18908 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12332 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12015 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15224 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12492 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10383 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11763 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11713 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3210 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16081 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12337 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -356,6 +356,13 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
+#if ENABLE(DARK_MODE_CSS)
+    if (layer() && oldStyle && oldStyle->colorScheme() != newStyle.colorScheme()) {
+        if (auto* scrollableArea = layer()->scrollableArea())
+            scrollableArea->invalidateScrollbars();
```
I tried also adding `scrollableArea->invalidateScrollCornerRect(scrollableArea->scrollCornerRect());` here but that didn't fix the scroll corner not updating.

pxlcoder replied:
May be worth looking into if `LocalFrameView::paintScrollCorner` gets called, and checking the value of `useDarkAppearanceForScrollbars()`.

lukewarlow replied:
I've found the fix for non-viewport scroll corners (when scrollbars are rendered). I needed invalidateScrollCorner rather than calling invalidateScrollCornerRect

pxlcoder commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -356,6 +356,13 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
     }
 
+#if ENABLE(DARK_MODE_CSS)
+    if (layer() && oldStyle && oldStyle->colorScheme() != newStyle.colorScheme()) {
```
I wonder if there are other things we need to update on a `color-scheme` change? Are system colors (used as a background / text color) updating correctly here?

This solution is fine, but I wonder if there's something existing we can hook into – I need to think a bit.

lukewarlow replied:
`data:text/html,<p>Hello</p><style>:root%20{color-scheme:light;}</style><script>setTimeout(()=>document.documentElement.style.colorScheme="dark",500);</script>`

From a quick test in Safari 16 using the above URL the system colours do seem to be updating.

Interestingly the viewport level scrollbars do update (excepting the scroll corner), so not sure why non viewport scrollers don't get invalidated correctly. Perhaps the viewport ones get invalidated as a side effect

pxlcoder replied:
Yeah, that test case will work because of the `recalculateBaseBackgroundColor` call for the root element.

Would be more interesting to change the `color-scheme` of a specific element.

lukewarlow replied:
`data:text/html,<div><p>Hello</p></div><style>:root%20{color-scheme:light;}div{color:CanvasText;background:Canvas}</style><script>setTimeout(()=>document.querySelector('div').style.colorScheme="dark",500);</script>`

This also seems to work. So I *think* it's just scrollbars specifically that aren't dynamic atm.

pxlcoder reviewed (commented):
Please retitle the PR title to describe the bug: "Scrollbar doesn't update on CSS color-scheme change" (taken from Bugzilla) is fine.

pxlcoder commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -371,6 +378,11 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
+        if (oldStyle && oldStyle->colorScheme() != newStyle.colorScheme()) {
+            view().frameView().invalidateScrollCorner(view().frameView().scrollCornerRect());
+            view().frameView().invalidateScrollbarsForAllScrollableAreas();
+        }
```
I think this won't work if you change the `color-scheme` of a specific-element? Since it's underneath `isDocElementRenderer || isBodyRenderer`.

lukewarlow replied:
Oh you might be right actually. I'll add the other line back to the above code snippet. Won't be at a computer till Thursday so apologies for slow response.

pxlcoder replied:
No worries! Appreciate your work on this 😄

pxlcoder reviewed (commented):
Nice tests! But I also think we'll want some test cases that deal with elements other than the document element.

pxlcoder commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -371,6 +380,9 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
         view().frameView().recalculateBaseBackgroundColor();
+
+        if (oldStyle && oldStyle->colorScheme() != newStyle.colorScheme())
+            view().frameView().invalidateScrollCorner(view().frameView().scrollCornerRect());
```
Do we also need to `invalidateScrollbars()` here? If so, it might be nice to factor:

```
scrollableArea->invalidateScrollCorner(scrollableArea->scrollCornerRect());
scrollableArea->invalidateScrollbars();
```

into a method.

pxlcoder replied:
Also, is this code necessary given the added logic above? The document element renderer should have a layer, but maybe the body does not? Want to make sure this isn't redundant with the earlier invalidation.

lukewarlow replied:
Viewport scrollbars already correctly invalidate (not sure where that happens perhaps as a side effect rather than a direct function call), so we don't need invalidateScrollbars for this block. 

We do however need the scroll corner invalidate because that's not handled elsewhere nor by the above code (I'm rebuilding locally just to double check this but I'm 99% sure)

pxlcoder replied:
I wonder if we need to make `LocalFrameView::invalidateScrollbarsForAllScrollableAreas()` also invalidate the scroll corner...

lukewarlow replied:
I'll take a look at what that will do. It might remove the neccessity to do the block for documents here.

lukewarlow replied:
It doesn't allow me to remove any of the code in RendeBox (Which is all neccessary) but it does fix the issue with  scroll corner changing when "color-scheme: light dark" is specified and the scroll corner doesn't repaint. So I've added that change will push it up now.

pxlcoder commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -356,6 +356,15 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
+    if (layer() && oldStyle && oldStyle->colorScheme() != newStyle.colorScheme()) {
+        if (auto* scrollableArea = layer()->scrollableArea()) {
+            scrollableArea->invalidateScrollCorner(scrollableArea->scrollCornerRect());
+            scrollableArea->invalidateScrollbars();
```
This logic is similar enough to the logic added in `LocalFrameView::invalidateScrollbarsForAllScrollableAreas`, which makes me wonder if we should just fold `scrollableArea->invalidateScrollCorner(scrollableArea->scrollCornerRect())` into `ScrollableArea::invalidateScrollbars`, rather than having it written out in both places.

The only other call to `invalidateScrollbars()` is in `ScrollableArea::setScrollbarOverlayStyle`, so I think it would be fine to share the code.

lukewarlow replied:
I'll do this locally and double check it works as expected (including the below comment)

pxlcoder commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -371,6 +380,9 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
         view().frameView().recalculateBaseBackgroundColor();
+
+        if (oldStyle && oldStyle->colorScheme() != newStyle.colorScheme())
+            view().frameView().invalidateScrollCorner(view().frameView().scrollCornerRect());
```
If you take my suggestion in https://github.com/WebKit/WebKit/pull/15264/files#r1268696498, I expect this code can be removed, as the `view().frameView().recalculateScrollbarOverlayStyle()` should take care of this.

lukewarlow committed:
Scrollbar doesn't update on CSS color-scheme change
https://bugs.webkit.org/show_bug.cgi?id=244494

Reviewed by Aditya Keerthi.

The color-scheme property is now diffed and triggers
invalidateScrollbars on scrollable areas.

invalidateScrollbars also now calls invalidateScrollCorners.

This causes the scrollbar and scroll corner to correctly rerender without
hovering or invalidating through other means. When UI Side Compositing is
disabled.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-1-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-1-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-1.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-2-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-2-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-2.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-3-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-3-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-3.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-4-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-4-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-scheme-dynamic-4.html: Added.
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::invalidateScrollbars):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::styleDidChange):

Canonical link: https://commits.webkit.org/266176@main
